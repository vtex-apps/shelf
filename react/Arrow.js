import React, { Component } from 'react'
import PropTypes from 'prop-types'

/**
 * Arrow component. It's an overrided component of react-slick that controls
 * the slide transition.
 */
class Arrow extends Component {
  render() {
    const { className, style, onClick, color } = this.props
    return (
      <div
        className={className}
        style={{ ...style, color }}
        onClick={onClick}
      />
    )
  }
}

Arrow.propTypes = {
  /** The css class of the element. */
  className: PropTypes.string,
  /** The custom style of the element. */
  style: PropTypes.object,
  /** The onClick handle function. */
  onClick: PropTypes.func,
  /** The color of the arrow icon. Ex: '#F00', 'rgb(255, 0, 0)'. */
  color: PropTypes.string.isRequired,
}

export default Arrow
