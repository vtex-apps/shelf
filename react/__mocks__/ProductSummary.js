import React, { Component } from 'react'

export default class ProductSummaryMock extends Component {
  render() {
    return (
      <div>Product Sumamry</div>
    )
  }
}
