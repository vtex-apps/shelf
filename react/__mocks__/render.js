import React, { Component } from 'react'

/**
 * NoSSR Mocked Component.
 */
export class NoSSR extends Component {
  render() {
    return (
      <div>{this.props.children}</div>
    )
  }
}