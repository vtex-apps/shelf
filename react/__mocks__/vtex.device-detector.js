export const useDevice = () => ({ isMobile: true })
