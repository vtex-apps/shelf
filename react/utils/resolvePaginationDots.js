export function resolvePaginationDotsVisibility(visibility, isMobile) {
  return !!(
    visibility === 'visible' ||
    (visibility === 'mobileOnly' && isMobile) ||
    (visibility === 'desktopOnly' && !isMobile)
  )
}
