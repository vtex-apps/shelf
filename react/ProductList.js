import './global.css'

import Spinner from '@vtex/styleguide/lib/Spinner'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import ProductSummary from 'vtex.product-summary/ProductSummary'

import VTEXClasses from './CustomClasses'
import { productListSchemaPropTypes } from './propTypes'
import ScrollTypes, { getScrollNames, getScrollValues } from './ScrollTypes'
import ShelfContent from './ShelfContent'

const DEFAULT_MAX_ITEMS = 10
const DEFAULT_ITEMS_PER_PAGE = 5

/**
 * Product List Component. Shows a collection of products.
 */
export default class ProductList extends Component {
  render() {
    const {
      products,
      loading,
      maxItems,
      titleText,
      arrows,
      scroll,
      itemsPerPage,
      summary,
    } = this.props
    return (
      <div className={`${VTEXClasses.MAIN_CLASS} ml7 mr7 pv4 pb7`}>
        <div
          className={`${
            VTEXClasses.TITLE_CONTENT_CLASS
          } w-100 flex justify-center`}>
          <h1 className={VTEXClasses.TITLE_TEXT_CLASS}> {titleText}</h1>
        </div>
        {loading ? (
          <div className="w-100 flex justify-center">
            <div className="w3 ma0">
              <Spinner />
            </div>
          </div>
        ) : (
          <ShelfContent
            products={products}
            maxItems={maxItems}
            arrows={arrows}
            scroll={scroll}
            itemsPerPage={itemsPerPage}
            summary={summary}
          />
        )}
      </div>
    )
  }
}

ProductList.getSchema = props => {
  return {
    title: 'editor.shelf.title',
    description: 'editor.shelf.description',
    type: 'object',
    properties: {
      maxItems: {
        title: 'editor.shelf.maxItems.title',
        type: 'number',
        default: ProductList.defaultProps.maxItems,
      },
      itemsPerPage: {
        title: 'editor.shelf.itemsPerPage.title',
        type: 'number',
        enum: [3, 4, 5],
        default: ProductList.defaultProps.itemsPerPage,
      },
      scroll: {
        title: 'editor.shelf.scrollType.title',
        type: 'string',
        enum: getScrollValues(),
        enumNames: getScrollNames(),
        default: ScrollTypes.BY_PAGE.value,
      },
      arrows: {
        title: 'editor.shelf.arrows.title',
        type: 'boolean',
        default: ProductList.defaultProps.arrows,
      },
      titleText: {
        title: 'editor.shelf.titleText.title',
        type: 'string',
        default: 'Default Title',
      },
      summary: {
        title: 'editor.shelf.summary.title',
        type: 'object',
        properties: ProductSummary.getSchema(props).properties,
      },
    },
  }
}

ProductList.defaultProps = {
  maxItems: DEFAULT_MAX_ITEMS,
  itemsPerPage: DEFAULT_ITEMS_PER_PAGE,
  scroll: ScrollTypes.BY_PAGE.value,
  arrows: true,
  titleText: 'Default Title',
}

ProductList.propTypes = {
  /** Loading status */
  loading: PropTypes.bool,
  /** Graphql data response. */
  products: ShelfContent.propTypes.products,
  ...productListSchemaPropTypes,
}
