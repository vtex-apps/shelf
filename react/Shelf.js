import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql } from 'react-apollo'

import Slider from 'react-slick'
import Arrow from './Arrow'
import Dots from './Dots'
import Spinner from '@vtex/styleguide/lib/Spinner'
import spinnerStyle from '@vtex/styleguide/lib/Spinner/style.css'

import productsQuery from './graphql/productsQuery.gql'

import { ExtensionPoint } from 'render'

/**
 * Shelf Component. Shows a collection of products.
 */
class Shelf extends Component {
  configureSettings() {
    const { arrows, scroll } = this.props

    return {
      slidesToShow: 5,
      slidesToScroll: scroll === 'BY_PAGE' ? 5 : 1,
      dots: true,
      arrows,
      nextArrow: <Arrow color="#000" />,
      prevArrow: <Arrow color="#000" />,
      infinite: false,
      appendDots: dots => <Dots color="#000" dots={dots} />,
      responsive: [
        {
          breakpoint: 1024,
          settings: {
            slidesToShow: 3,
            slidesToScroll: scroll === 'BY_PAGE' ? 3 : 1,
          },
        },
        {
          breakpoint: 600,
          settings: {
            slidesToShow: 1,
            slidesToScroll: 1,
          },
        },
      ],
    }
  }

  normalizeProductSummaryProps(product) {
    return {
      listPrice: product.items[0].sellers[0].commertialOffer.Price,
      sellingPrice: product.items[0].sellers[0].commertialOffer.Price,
      imageUrl: product.items[0].images[0].imageUrl.replace('http:', ''),
      url: `/product/${product.productId}`,
      name: product.productName,
    }
  }

  isEditMode() {
    return !!document.querySelector('.edit-mode')
  }

  renderSlideProperly(products, maxItems) {
    const slideSettings = this.configureSettings()
    if (this.isEditMode()) {
      if (products.length) {
        return (
          <div className="w-20" key={products[0].productId}>
            <ExtensionPoint id="shelfitem"
              product={this.normalizeProductSummaryProps(products[0])}>
            </ExtensionPoint>
          </div>
        )
      }
      return (
        <div className="w-20" key="1">
          <ExtensionPoint id="shelfitem"
            product={{
              listPrice: 200,
              sellingPrice: 40,
              imageUrl: '//www.allfree-clipart.com/Business/computer.jpg',
              url: '/product/1',
              name: 'Product1',
            }}>
          </ExtensionPoint>
        </div>
      )
    }
    return (
      <Slider {...slideSettings}>
        {
          products.slice(0, maxItems).map(item => {
            return (
              <div key={item.productId}>
                <ExtensionPoint id={'shelfitem'}
                  product={this.normalizeProductSummaryProps(item)}>
                </ExtensionPoint>
              </div>
            )
          })
        }
      </Slider>
    )
  }

  render() {
    const { data, maxItems, titleText } = this.props
    const products = !data || data['error'] ? [] : data.products

    return (
      <div className="ml7 mr7 pv4 vtex-shelf">
        <div className="w-100 flex justify-center">
          <h1> {titleText}</h1>
        </div>
        {
          data.loading && (
            <div className="w-100 flex justify-center">
              <div className="w3 ma0">
                <Spinner style={spinnerStyle} />
              </div>
            </div>
          )
        }
        {
          !data.loading && products && (
            this.renderSlideProperly(products, maxItems)
          )
        }
      </div>
    )
  }
}

Shelf.schema = {
  title: 'Shelf',
  description: 'A product shelf featuring a collection',
  type: 'object',
  properties: {
    category: {
      title: 'Category',
      type: 'number',
    },
    collection: {
      title: 'Collection',
      type: 'number',
    },
    orderBy: {
      title: 'List Ordenation',
      type: 'string',
      enum: ['OrderByTopSaleDESC', 'OrderByPriceDESC', 'OrderByPriceASC'],
      enumNames: ['Sales', 'Price, descending', 'Price, ascending'],
      default: 'OrderByTopSaleDESC',
    },
    maxItems: {
      title: 'Max Items',
      type: 'number',
      default: 10,
    },
    scroll: {
      title: 'Scroll Type',
      type: 'string',
      enum: ['BY_PAGE', 'ONE_TO_ONE'],
      enumNames: ['By Page', 'One To One'],
      default: 'BY_PAGE',
    },
    arrows: {
      title: 'Arrows',
      type: 'boolean',
      default: true,
    },
    titleText: {
      title: 'Title Text',
      type: 'string',
      default: 'Default Title',
    },
  },
}

Shelf.defaultProps = {
  maxItems: 10,
  scroll: 'BY_PAGE',
  arrows: true,
}

Shelf.propTypes = {
  /** The graphql data response. */
  data: PropTypes.shape({
    products: PropTypes.arrayOf(PropTypes.shape({
      productId: PropTypes.string.isRequired,
      productName: PropTypes.string.isRequired,
      items: PropTypes.arrayOf(PropTypes.shape({
        images: PropTypes.arrayOf(PropTypes.shape({
          imageUrl: PropTypes.string.isRequired,
        })).isRequired,
        sellers: PropTypes.arrayOf(PropTypes.shape({
          commertialOffer: PropTypes.shape({
            Price: PropTypes.number.isRequired,
          }),
        })).isRequired,
      })),
    })),
  }).isRequired,
  /** The Category Id. */
  category: PropTypes.number,
  /** The Collection Id. */
  collection: PropTypes.number,
  /** The Ordenation Type. */
  orderBy: PropTypes.string,
  /** Maximum number of items in the shelf. */
  maxItems: PropTypes.number.isRequired,
  /** The scroll options. */
  scroll: PropTypes.string.isRequired,
  /** The Collection Id. */
  arrows: PropTypes.bool.isRequired,
  /** The text value of the title. */
  titleText: PropTypes.string,
}

const options = {
  options: ({
    category = undefined,
    orderBy = undefined,
    maxItems = 10,
    collection = undefined,
  }) => ({
    variables: {
      category: category || undefined,
      collection: collection || undefined,
      specificationFilters: [],
      orderBy: orderBy || undefined,
      from: 0,
      to: maxItems - 1,
    },
    ssr: false,
  }),
}

export default graphql(productsQuery, options)(Shelf)
