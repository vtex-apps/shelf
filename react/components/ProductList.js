import PropTypes from 'prop-types'
import { identity, path } from 'ramda'
import React, { Component, Fragment } from 'react'
import ReactResizeDetector from 'react-resize-detector'
import { IOMessage } from 'vtex.native-types'

import { productListSchemaPropTypes, shelfItemPropTypes } from '../utils/propTypes'
import ScrollTypes, { getScrollNames, getScrollValues } from '../utils/ScrollTypes'
import GapPaddingTypes, {
  getGapPaddingNames,
  getGapPaddingValues,
} from '../utils/paddingEnum'
import ShelfContent from './ShelfContent'

import shelf from './shelf.css'

const DEFAULT_MAX_ITEMS = 10
const DEFAULT_ITEMS_PER_PAGE = 5

function normalizeBuyable(product) {
  const items = path(['items'], product)
  const buyableItems =
    path(['length'], items) &&
    items
      .map(item => ({
        ...item,
        sellers: getBuyableSellers(item.sellers),
      }))
      .filter(item => path(['sellers', 'length'], item))

  return buyableItems
    ? {
        ...product,
        items: buyableItems,
      }
    : null
}

function getBuyableSellers(sellers) {
  return (
    path(['length'], sellers) &&
    sellers.filter(seller => path(['sellerId'], seller))
  )
}

/**
 * Product List Component. Shows a collection of products.
 */
class ProductList extends Component {
  render() {
    const {
      products,
      maxItems,
      titleText,
      arrows,
      scroll,
      itemsPerPage,
      summary,
      isMobile,
      gap,
      showTitle,
    } = this.props

    const filteredProducts =
      products && products.map(normalizeBuyable).filter(identity)

    return products && !products.length ? null : (
      <Fragment>
        {showTitle && (
          <div
            className={`${
              shelf.title
            } t-heading-2 fw3 w-100 flex justify-center pt7 pb6 c-muted-1`}
          >
            <IOMessage id={titleText} />
          </div>
        )}
        <ReactResizeDetector handleWidth>
          {width => (
            <ShelfContent
              products={filteredProducts}
              maxItems={maxItems}
              arrows={arrows}
              scroll={scroll}
              itemsPerPage={itemsPerPage}
              summary={summary}
              isMobile={isMobile}
              width={width}
              gap={gap}
            />
          )}
        </ReactResizeDetector>
      </Fragment>
    )
  }
}

ProductList.getSchema = props => {
  return {
    title: 'admin/editor.shelf.title',
    description: 'admin/editor.shelf.description',
    type: 'object',
    properties: {
      maxItems: {
        title: 'admin/editor.shelf.maxItems.title',
        type: 'number',
        default: ProductList.defaultProps.maxItems,
        isLayout: true,
      },
      gap: {
        title: 'admin/editor.shelf.gap.title',
        type: 'string',
        enum: getGapPaddingValues(),
        enumNames: getGapPaddingNames(),
        default: GapPaddingTypes.SMALL.value,
        isLayout: true,
      },
      itemsPerPage: {
        title: 'admin/editor.shelf.itemsPerPage.title',
        type: 'number',
        enum: [3, 4, 5],
        default: ProductList.defaultProps.itemsPerPage,
        isLayout: true,
      },
      scroll: {
        title: 'admin/editor.shelf.scrollType.title',
        type: 'string',
        enum: getScrollValues(),
        enumNames: getScrollNames(),
        default: ScrollTypes.BY_PAGE.value,
        isLayout: true,
      },
      arrows: {
        title: 'admin/editor.shelf.arrows.title',
        type: 'boolean',
        default: ProductList.defaultProps.arrows,
        isLayout: true,
      },
      showTitle: {
        title: 'admin/editor.shelf.titleText.showTitle',
        type: 'boolean',
        default: ProductList.defaultProps.showTitle,
        isLayout: true,
      }
    },
  }
}

ProductList.defaultProps = {
  maxItems: DEFAULT_MAX_ITEMS,
  itemsPerPage: DEFAULT_ITEMS_PER_PAGE,
  scroll: ScrollTypes.BY_PAGE.value,
  gap: GapPaddingTypes.SMALL.value,
  arrows: true,
  showTitle: true,
  titleText: null,
  isMobile: false,
}

ProductList.propTypes = {
  /** Loading status */
  loading: PropTypes.bool,
  /** Graphql data response. */
  products: PropTypes.arrayOf(shelfItemPropTypes.item),
  /** Verifies if is a mobile device. */
  isMobile: PropTypes.bool,
  ...productListSchemaPropTypes,
}

export default ProductList
