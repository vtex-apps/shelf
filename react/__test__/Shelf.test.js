/* eslint-env jest */
import React from 'react'
import { render } from '@vtex/test-tools/react'
import ProductList from '../components/ProductList'
import { productMock } from '../__mocks__/productMock'
import Shelf from '../Shelf'
import { resolvePaginationDots } from '../utils/resolvePaginationDots'
const fs = require('fs')

describe('Shelf component', () => {
  const renderComponent = customProps => {
    const props = {
      maxItems: 1,
      itemsPerPage: 1,
      arrows: true,
      showTitle: true,
    }

    // TODO: Remove this later when we have a better way to resolve contentSchemas.json at test-tools
    const rawData = fs.readFileSync('../store/contentSchemas.json')
    const contentSchema = JSON.parse(rawData)
    const titleTextId = contentSchema.definitions.ProductList.properties.titleText.default

    const wrapper = render(<ProductList titleText={titleTextId} {...props} {...customProps} />)
    return wrapper
  }

  it('should be rendered', () => {
    const component = renderComponent()
    expect(component).toBeDefined()
  })

  it('should render nothing if there is no products', () => {
    const component = renderComponent({ products: [] })

    expect(component.container.querySelector('.title')).toBeFalsy()
  })

  it('should match the snapshot', () => {
    const component = renderComponent({ products: productMock })
    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should show in Site Editor', () => {
    const schema = Shelf.schema || Shelf.getSchema({})
    expect(schema).toBeDefined()
    expect(typeof schema.title).toBe('string')
  })
})

describe('Util functions', () => {
  it('should return the correct value for pagination dots visibility', () => {
    const MOBILE_ONLY = 'mobileOnly'
    const DESKTOP_ONLY = 'desktopOnly'
    const VISIBLE = 'visible'
    const HIDDEN = 'hidden'

    /** resolvePaginationDots(visibility: string, isMobile: boolean) */
    expect(resolvePaginationDots(MOBILE_ONLY, false)).toBeFalsy()
    expect(resolvePaginationDots(MOBILE_ONLY, true)).toBeTruthy()

    expect(resolvePaginationDots(DESKTOP_ONLY, false)).toBeTruthy()
    expect(resolvePaginationDots(DESKTOP_ONLY, true)).toBeFalsy()

    expect(resolvePaginationDots(VISIBLE, false)).toBeTruthy()
    expect(resolvePaginationDots(VISIBLE, true)).toBeTruthy()

    expect(resolvePaginationDots(HIDDEN, false)).toBeFalsy()
    expect(resolvePaginationDots(HIDDEN, true)).toBeFalsy()
  })
})
