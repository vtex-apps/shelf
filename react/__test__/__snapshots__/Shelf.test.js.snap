// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shelf component should match snapshot 1`] = `
<div>
  <div
    class="vtex-shelf ml7 mr7 pv4"
  >
    <div
      class="vtex-shelf__title-content w-100 flex justify-center"
    >
      <h1
        class="vtex-shelf__title-text"
      >
         
        Default Title
      </h1>
    </div>
    <div>
      <div>
        Slider
      </div>
    </div>
  </div>
</div>
`;
